File: day3/part1.py
def house_visited(directions):
    visited = 1
    visited_list = [[0, 0]]
    robo_coords = [0, 0]
    robo_coords = [0, 0]

    for char in directions:
        if char == "^":
            new_coords = [robo_coords[0], robo_coords[1]+1]
        elif char == "v":
            new_coords = [robo_coords[0], robo_coords[1]-1]
        elif char == "<":
            new_coords = [robo_coords[0]-1, robo_coords[1]]
        elif char == ">":
            new_coords = [robo_coords[0]+1, robo_coords[1]]

        if new_coords not in visited_list:
            visited_list.append(new_coords)
            visited += 1

        robo_coords = new_coords

    return visited


def main():
    with open('input.txt', 'r') as f:
        p = f.read()

    print(house_visited(p))


main()

File: day3/part1.go
package main

import (
	"fmt"

	"github.com/sagnikc395/aoc2015/utils"
)

type Direction struct {
	X int
	Y int
}

func (d *Direction) UpdateDirection(direction string) {
	switch direction {
	case ">":
		d.X += 1
	case "^":
		d.Y += 1
	case "<":
		d.X -= 1
	case "v":
		d.Y -= 1
	}

}

func main() {
	f := utils.ReadFile("./input.txt")

	data := string(f)

	//start posn at 0,0
	var dir Direction

	visitedHouses := make(map[Direction]bool)

	//marking the starting house as visited
	visitedHouses[dir] = true

	for _, val := range data {
		if val == '>' {
			dir.UpdateDirection(">")
		} else if val == '<' {
			dir.UpdateDirection("<")
		} else if val == 'v' {
			dir.UpdateDirection("v")
		} else if val == '^' {
			dir.UpdateDirection("^")
		}
		visitedHouses[dir] = true
	}
	fmt.Printf("Number of houses that received at least one present: %d\n", len(visitedHouses))
	fmt.Printf("Final Position (%d ,%d)", dir.X, dir.Y)
}
